src/ringmesh/mesh/volume_mesh.cpp

```diff
@@ -0,0 +1,351 @@
+                nb_cell_facets() );
+            index_t cf = 0;
+            for( auto c : range( nb_cells() ) )
+            {
+                for( auto f : range( nb_cell_facets( c ) ) )
+                {
+                    cell_facet_centers[cf] = cell_facet_barycenter( { c, f } );
+                    ++cf;
+                }
+            }
+            cell_facet_nn_search_.reset(
+                new NNSearch< DIMENSION >( cell_facet_centers, true ) );
+        }
+        return *cell_facet_nn_search_.get();
+    }
+
+    template < index_t DIMENSION >
+    std::tuple< index_t, std::vector< index_t > >
+        VolumeMesh< DIMENSION >::connected_components() const
+    {
+        std::vector< index_t > components( nb_cells(), NO_ID );
+        index_t nb_components{ 0 };
+        for( auto cell : range( nb_cells() ) )
+        {
+            if( components[cell] == NO_ID )
+            {
+                std::stack< index_t > S;
+                S.push( cell );
+                components[cell] = nb_components;
+                do
+                {
+                    auto cur_cell = S.top();
+                    S.pop();
+                    for( auto facet : range( nb_cell_facets( cur_cell ) ) )
+                    {
+                        auto adj_cell = cell_adjacent( { cur_cell, facet } );
+                        if( adj_cell != NO_ID && components[adj_cell] == NO_ID )
+                        {
+                            S.push( adj_cell );
+                            components[adj_cell] = nb_components;
+                        }
+                    }
+                } while( !S.empty() );
+                nb_components++;
+            }
+        }
+        return std::make_tuple( nb_components, components );
+    }
+
+    template < index_t DIMENSION >
+    bool VolumeMesh< DIMENSION >::is_mesh_valid() const
+    {
+        bool valid{ true };
+
+        if( this->nb_vertices() < 4 )
+        {
+            Logger::warn( "VolumeMesh has less than 4 vertices " );
+            valid = false;
+        }
+        if( nb_cells() == 0 )
+        {
+            Logger::warn( "VolumeMesh has no cell" );
+            valid = false;
+        }
+
+        // No isolated vertices
+        std::vector< index_t > nb( this->nb_vertices(), 0 );
+        for( auto c : range( nb_cells() ) )
+        {
+            for( auto v : range( nb_cell_vertices( c ) ) )
+            {
+                nb[cell_vertex( { c, v } )]++;
+            }
+        }
+        auto nb_isolated_vertices =
+            static_cast< index_t >( std::count( nb.begin(), nb.end(), 0 ) );
+        if( nb_isolated_vertices > 0 )
+        {
+            Logger::warn( "VolumeMesh", "Mesh has ", nb_isolated_vertices,
+                " isolated vertices " );
+            valid = false;
+        }
+
+        return valid;
+    }
+
+    template < index_t DIMENSION >
+    std::vector< index_t > VolumeMesh< DIMENSION >::cells_around_vertex(

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 26 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3776)

Reply:
Je n'ai pas modifié cette fonction. Dois-je la factoriser?